src/e3response/data/qm9_nmr.py

```diff
@@ -0,0 +1,455 @@
+            self.dataset = [dataset]
+        else:
+            self.dataset = list(dataset)
+
+        for ds in self.dataset:
+            if ds not in DATASET_URLS:
+                raise ValueError(
+                    f"Dataset '{ds}' not recognised. Available: {list(DATASET_URLS.keys())}"
+                )
+
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir, exist_ok=True)
+
+        # Params
+        self._rmax = r_max
+        self._data_dir: Final[str] = data_dir
+        self._limit = limit
+        default_key = [
+            "NMR_tensors",
+        ]
+        possible_keys = [
+            "ind",
+            "N",
+            "species",
+            "isotropic",
+            "anisotropy",
+            "eigenvalues",
+        ]
+
+        if isinstance(atom_keys, str):
+            atom_keys = [atom_keys]
+
+        invalid_keys = [key for key in (atom_keys or []) if key not in possible_keys]
+        if invalid_keys:
+            raise ValueError(
+                f"Invalid atom_keys: {invalid_keys}. " f"Allowed keys are: {possible_keys}"
+            )
+
+        self._atom_keys = list(set(default_key).union(atom_keys or []))
+
+        self._to_graph: Callable[[ase.Atoms], jraph.GraphsTuple] = functools.partial(
+            gcnn.atomic.graph_from_ase,
+            r_max=self._rmax,
+            atom_include_keys=self._atom_keys,
+        )
+
+        # Data
+        self._data = []
+        for ds in self.dataset:
+            archive_name = f"QM9nmr_{ds}_logs.zip"
+            archive_path = os.path.join(data_dir, archive_name)
+            url = DATASET_URLS[ds]
+
+            if os.path.isfile(archive_path):
+                try:
+                    with zipfile.ZipFile(archive_path, "r") as zip_ref:
+                        zip_ref.testzip()
+                except (zipfile.BadZipFile, zipfile.LargeZipFile, IOError) as e:
+                    _LOGGER.warning(
+                        f"{archive_name} is corrupted or unreadable: {e}, removing corrupted archive ..."
+                    )
+                    os.remove(archive_path)
+                    self._download_file(archive_name, url, archive_path)
+                else:
+                    _LOGGER.info(f"{archive_name} already present and valid at {archive_path}")
+            else:
+                _LOGGER.info(f"{archive_name} not found.")
+                self._download_file(url, archive_path)
+
+            structures = self._extract_archive_zip(archive_path, limit=self._limit)
+            self._data.extend(structures)
+
+    def __getitem__(self, index):
+        return self._to_graph(self._data[index])  # translation in graphs on demand
+
+    def __len__(self):
+        return len(self._data)
+
+    def _download_file(self, name: str, url: str, path: str) -> None:
+        _LOGGER.info(f"\nDownloading {name} from {url} ...")
+        try:
+            with urllib.request.urlopen(url) as response:  # nosec B310
```

Comment:
Can you change this to use the `reporthook` parameter of `requests.urlretrieve()`.  This is more elegant, and means that you don't have to write the 'chunking' code yourself, which is error prone.  See:

https://pypi.org/project/tqdm/3.4.0/#hooks-and-callbacks

for an example